### PR TITLE
Text-to-Speech

### PR DESCRIPTION
**Pull request by Hamms**
Currently, just a button in the instructions that lets the user select from one of three voices to read the instructions aloud to them.

image

### CODE REVIEW

**Comment by bcjordan** [comment]
@Hamms super cool :)

**Comment by Hamms** [comment]
Check out https://dashboard-adhoc-tts.cdn-code.org/s/allthethings/stage/4/puzzle/4 for the live demo!

This is a WIP because we are currently evaluating the best way to integrate a TTS service into our workflow so we can generate audio content at deploy time and serve over a CDN, rather than doing what we're currently doing and making a request against an external TTS service every time "play" is pressed

**Comment by bcjordan** [comment]
Hahah the Yodit voice :). This is really cool!

**Comment by Hamms** [comment]
Please feel free to test [on our adhoc server](https://dashboard-adhoc-tts.cdn-code.org/s/course1?enableExperiments=tts).

Note that we are preparing to conduct usability tests on two candidate voices, which you can toggle between by adding the url parameter `?voice=a` or `?voice=b`.

File: dashboard/app/views/levels/editors/_tts.haml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    = f.label :voice
+    %div
+      %select#level_tts_voice
+    %div
+      %button.tts#tts-instructions{type: 'button'} Hear Instructions
+      %button.tts#tts-markdown-instructions{type: 'button'} Hear Markdown Instructions
+    %div
+      %audio#tts-audio
+
+:javascript
```
Comment (joshlory) [inline dashboard/app/views/levels/editors/_tts.haml]: Extract to a standalone JavaScript file.

File: dashboard/app/models/concerns/text_to_speech.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    )
+  end
+
+  def self.tts_upload_to_s3(text, filename, voice=:rosie)
+    return if text.blank?
+    return if AWS::S3.exists_in_bucket(TTS_BUCKET, filename)
+    url = acapela_text_to_audio_url(text, VOICES[voice][:VOICE], VOICES[voice][:SPEED], VOICES[voice][:SHAPE])
+    return if url.nil?
+    uri = URI.parse(url)
+    Net::HTTP.start(uri.host) { |http|
```
Comment (joshlory) [inline dashboard/app/models/concerns/text_to_speech.rb]: Super small nit: [prefer `do...end` for multiline blocks](https://github.com/bbatsov/ruby-style-guide#single-line-blocks).

File: dashboard/app/helpers/levels_helper.rb [hunk 3]
```diff
@@ -354,6 +354,19 @@ def blockly_options
       app_options['pusherApplicationKey'] = CDO.pusher_application_key
     end
 
+    if CDO.use_acapela
+      # TTS is currently only enabled for k1
+      app_options['acapelaEnabled'] = CDO.use_acapela && script.is_k1?
+      # temporary parameter for a/b testing our voice options
```
Comment (joshlory) [inline dashboard/app/helpers/levels_helper.rb]: TODO.

File: dashboard/app/models/concerns/text_to_speech.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    SPEED: 180,
+    SHAPE: 100
+  }
+}
+
+module TextToSpeech
+  extend ActiveSupport::Concern
+
+  # TODO: this concern actually depends on the SerializedProperties
+  # concern ... I'm not sure how best to deal with that.
```
Comment (joshlory) [inline dashboard/app/models/concerns/text_to_speech.rb:34]: TODO.

**Comment by joshlory** [comment]
It would be great to have test coverage for this feature!

File: dashboard/app/views/levels/editors/_tts.haml (hunk 1 above)
Comment (Hamms) [inline dashboard/app/views/levels/editors/_tts.haml]: Where's the preferred place to put that JS file these days?

File: dashboard/app/models/concerns/text_to_speech.rb (hunk 4 above)
Comment (Hamms) [inline dashboard/app/models/concerns/text_to_speech.rb:34]: Any thoughts on the appropriate way to deal with this? Should this concern include the SerializedProperties concern? Or perhaps extend it?

File: dashboard/app/views/levels/editors/_tts.haml (hunk 1 above)
Comment (joshlory) [inline dashboard/app/views/levels/editors/_tts.haml]: `/apps/src/code-studio` is probably the best place.

**Comment by joshlory** [comment]
Sorry, I forgot this was on my plate.  LGTM after moving the JS to `/apps`.

**Comment by Hamms** [comment]
PTA(quick)L at the new commits

File: apps/src/code-studio/acapela.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import $ from 'jquery';
+
+module.exports = {
+  getAudio: function (text, voice, config, callback) {
+    $.getJSON("http://vaas.acapela-group.com/Services/UrlMaker?jsoncallback=?", {
```
Comment (joshlory) [inline apps/src/code-studio/acapela.js]: What is the `jsoncallback=?`?

File: apps/src/code-studio/acapela.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      cl_login: config.login,
+      cl_app: config.app,
+      cl_pwd: config.pwd,
+      req_voice: voice,
+      req_text: text
+    }, callback);
+  },
+
+  populateVoicesList: function (target, config) {
```
Comment (joshlory) [inline apps/src/code-studio/acapela.js]: Super small nit: could be `populateVoicesList(target, config) {`.

File: apps/src/code-studio/acapela.js (hunk 5 above)
Comment (Hamms) [inline apps/src/code-studio/acapela.js]: just a param that acapela expects when making this request with JSON

**Comment by joshlory** [comment]
LGTM! 👍

**Pull request merged**